### ISSUE TITLE
Helm buffer matching no longer matches erc buffers when not starting with a #

### ISSUE THREAD

**Issue by TLATER** [emacs-helm/helm#2224]
## Expected behavior
I believe something about helm's buffer matching changed, presumably deliberately because of buffer naming conventions, but this breaks my workflow with erc.

When selecting buffers from a collection like:

```
#helm
#emacs
#other-irc-channel
```
I would like to be able to find the buffer with its name, ommitting the #.

## Actual behavior (from `emacs-helm.sh` if possible, see note at the bottom)

Helm will not display any buffers when searching, unless I start the buffer name with a #. This changed recently, perhaps over the last week.

## Steps to reproduce (recipe)

1. Create a buffer like `#helm`
2. Use helm for `switch-to-buffer`
3. Type `helm`
4. See that no buffers are displayed

## Describe versions of Helm, Emacs, operating system, etc.

Nixos: 19.09
Emacs: 26.3
Helm: 20191112.2014 (melpa)

## Are you using `emacs-helm.sh` to reproduce this bug? (yes/no):

No, I don't think it's relevant per-se. Presumably a default changed, and I can't find what the default was nor when it changed. This is my full helm configuration, for reference:

```lisp
(use-package helm
             :config
             (require 'helm-config)
             (helm-mode 1)
             :custom
             (helm-ff-lynx-style-map t)
             :bind
             ([remap execute-extended-command] . helm-M-x)
             ([remap find-file] . helm-find-files)
             ("C-c i" . helm-imenu)
             ("C-c s" . helm-occur))
```

## Are you using Spacemacs? (yes/no):

no

**Comment by thierryvolpiatto** [comment]
Tristan Maat <notifications@github.com> writes:

> I believe something about helm's buffer matching changed,

A lot of things are changing actually.

> Steps to reproduce (recipe)
>
>  1. Create a buffer like #helm
>  2. Use helm for switch-to-buffer
>  3. Type helm
>  4. See that no buffers are displayed

Can't reproduce, please use emacs-helm.sh to reproduce your bug.
Couldn't reproduce with emacs 26.3 and 27.

-- 
Thierry

Get my Gnupg key:
gpg --keyserver pgp.mit.edu --recv-keys 59F29997

**Comment by akater** [comment]
I can't reproduce the non-matched `#helm` buffer by `helm` substring but I experience something similar: since recently, my `password-store-copy` entries of the form `dir/subdir/file` do not match by `file` anymore (or `subdir` for that matter) but they used to, and the completion has thus become much less useful. I'll try with emacs-helm.sh when I figure it out. I'm using Emacs 27.0.50 (master branch).

**Comment by thierryvolpiatto** [comment]
akater <notifications@github.com> writes:

> I can't reproduce the non-matched #helm buffer by helm substring but I
> experience something similar: since recently, my password-store-copy
> entries of the form dir/ subdir/file do not match by file anymore (or
> subdir for that matter) but they used to, and the completion has thus
> become much less useful. I'll try with emacs-helm.sh when I figure it
> out. I'm using Emacs 27.0.50 (master branch).

I doubt you will have better results as from what I can see after
downloding `password-store.el`, it uses `completing-read` to complete
against file names or directory names, `completing-read` doesn't know
about file paths, it matches whole filenames as a simple string, I
suggest password-store use the proper tool to complete against
file/directory names, that is `read-file-name` and/or
`read-directory-name`.  Also I don't see how it could be better before
last changes as long as `completing-read` is used for file names.

--
Thierry

Get my Gnupg key:
gpg --keyserver pgp.mit.edu --recv-keys 59F29997

**Comment by akater** [comment]
The last commit to emacs-password-store was in 2014. It was working fine for me for many years.

It does not make any sense to use `read-file-name` for elements of password-store: there is no need to make user traverse directory tree to pick a password. I used to pick my github password by typing github. Now I have to type some more of the complete path. This is clearly suboptimal.

Another example: I'm using completing-read to pick a gentoo package. Packages are [roughly] placed in a two-level directory tree: categories, then packages, i.e. `app-misc/beancounter`. It is broken now as well, and it's totally inconvenient to make users pick a category first: one might not even remember the relevant category (was it app-misc, or app-office like one similar package is categorized, or maybe some other category entirely?). In both these cases users don't pick files, they pick other entities that happen to be (imprecisely in both cases, in fact) represented by files in a directory tree.

**Comment by TLATER** [comment]
@thierryvolpiatto I can reproduce it running `emacs-helm.sh`

**Comment by thierryvolpiatto** [comment]
akater <notifications@github.com> writes:

> The last commit to emacs-password-store was in 2014. It was working
> fine for me for many years.
>
> It does not make any sense to use read-file-name for elements of
> password-store: there is no need to make user traverse directory tree
> to pick a password. I used to pick my github password by typing
> github. Now I have to type some more of the complete path. This is
> clearly suboptimal.
>
> Another example: I'm using completing-read to pick a gentoo
> package. Packages are [roughly] placed in a two-level directory tree:
> categories, then packages, i.e.  app-misc/beancounter. It is broken
> now as well, and it's totally inconvenient to make users pick a
> category first: one might not even remember the relevant category (was
> it app-misc, or app-office like one similar package is categorized, or
> maybe some other category entirely?). In both these cases users don't
> pick files, they pick other entities that happen to be (imprecisely in
> both cases, in fact) represented by files in a directory tree.

Please give me concrete recipe of what is working and not working for
you, I am not using gentoo nor password-store.
If you are not happy with what you have actually, try to set
helm-completion-style to 'helm, or try to play with
helm-completing-read-handlers-alist, you will have the previous
behavior, or use a stable version of helm.

--
Thierry

Get my Gnupg key:
gpg --keyserver pgp.mit.edu --recv-keys 59F29997

**Comment by akater** [comment]
Setting `helm-completion-style` to `helm` alleviates the issue with matching subexpressions after slashes, thank you @thierryvolpiatto. Hopefully tweaking this variable will be enough for @TLATER as well.

**Comment by TLATER** [comment]
That's... Interesting. Another issue mentioned more specific variables to set the completion style for specific kinds of prompts, one of which being the buffer prompt, and changing that to `'helm` didn't change anything.

I can't find those variables in `apropos` anymore, I suppose perhaps I caught an emacs instance where those variables were still set beforehand.

In either case, yes, `helm-completion-style` was what I was looking for, although I didn't think so previously, and setting that to `'emacs` by default seems reasonable - it would be nice if these kinds of changes to defaults could be communicated a little better though, second time I've ended up diving into issues for that :)

**Comment by thierryvolpiatto** [comment]
Tristan Maat <notifications@github.com> writes:

> @thierryvolpiatto I can reproduce it running emacs-helm.sh

Hmm, I see it was because you didn't add a space after "helm" and you
used only helm style (which is enabled by default when you start
completing), but yes it is legitimate to expect to have helm popping up
after entering "helm", should now be fixed.

-- 
Thierry

Get my Gnupg key:
gpg --keyserver pgp.mit.edu --recv-keys 59F29997

**Comment by thierryvolpiatto** [comment]
akater <notifications@github.com> writes:

> Setting helm-completion-style to helm alleviates the issue with
> matching subexpressions after slashes, thank you
> @thierryvolpiatto. Hopefully tweaking this variable will be enough for
> @TLATER as well.

I just commited a change that fix the default as well, so you probably
don't have to set helm-completion-style to 'helm, it should work
properly with 'emacs as well.

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.*


--
Thierry

Get my Gnupg key:
gpg --keyserver pgp.mit.edu --recv-keys 59F29997

**Comment by thierryvolpiatto** [comment]
Tristan Maat <notifications@github.com> writes:

> That's... Interesting. Another issue mentioned more specific variables
> to set the completion style for specific kinds of prompts, one of
> which being the buffer prompt, and changing that to 'helm didn't
> change anything.

No you don't have to add 'helm to `completion-styles`, it is added
automatically.

> I can't find those variables in apropos anymore, I suppose perhaps I
> caught an emacs instance where those variables were still set
> beforehand.

Which variables?

> In either case, yes, helm-completion-style was what I was looking for,
> although I didn't think so previously, and setting that to 'emacs by
> default seems reasonable -

Should work now with default value 'emacs, you don't have to switch to
'helm style.

> it would be nice if these kinds of changes to defaults could be
> communicated a little better though, second time I've ended up diving
> into issues for that :)

Helm completion is actually in heavy development, working in parallel
with emacs-27 development of completion API, so please be patient,
everything will be more clear after new helm release and as well after
emacs-27 release.

Thanks.

--
Thierry

Get my Gnupg key:
gpg --keyserver pgp.mit.edu --recv-keys 59F29997

**Comment by akater** [comment]
@thierryvolpiatto I confirm: recent helm update does indeed make post-slash matches fire the way they used to even with `helm-completion-style` set to `emacs`. Thank you!
